x-pack/plugins/lens/public/visualizations/xy/state_helpers.ts

```diff
@@ -117,65 +141,182 @@ function getLayerReferenceName(layerId: string) {
+          return persistedLayer as XYLayerConfig;
+        }
+
+        const indexPatternIdFromReferences =
+          references.find(({ name }) => name === getLayerReferenceName(persistedLayer.layerId))
+            ?.id || fallbackIndexPatternId;
+
+        let injectedLayer: XYAnnotationLayerConfig;
+
+        if (isPersistedByValueAnnotationsLayer(persistedLayer)) {
+          injectedLayer = {
+            ...persistedLayer,
+            indexPatternId:
+              // getIndexPatternIdFromInitialContext(persistedLayer, initialContext) || TODO - was this doing anything?
+              indexPatternIdFromReferences,
+          };
+        } else {
+          const annotationGroupId = references?.find(
+            ({ name }) => name === persistedLayer.annotationGroupRef
+          )?.id;
+
+          const annotationGroup = annotationGroupId
+            ? annotationGroups[annotationGroupId]
+            : undefined;
+
+          if (!annotationGroupId || !annotationGroup) {
+            return undefined; // the annotation group this layer was referencing is gone, so remove the layer
+          }
+
+          // declared as a separate variable for type checking
+          const commonProps: Pick<
+            XYByReferenceAnnotationLayerConfig,
+            'layerId' | 'layerType' | 'annotationGroupId' | '__lastSaved'
+          > = {
+            layerId: persistedLayer.layerId,
+            layerType: persistedLayer.layerType,
+            annotationGroupId,
+            __lastSaved: annotationGroup,
+          };
+
+          if (isPersistedByReferenceAnnotationsLayer(persistedLayer)) {
+            // a clean by-reference layer inherits everything from the library annotation group
+            injectedLayer = {
+              ...commonProps,
+              ignoreGlobalFilters: annotationGroup.ignoreGlobalFilters,
+              indexPatternId: annotationGroup.indexPatternId,
+              annotations: cloneDeep(annotationGroup.annotations),
+            };
+          } else {
+            // a linked by-value layer gets settings from visualization state while
+            // still maintaining the reference to the library annotation group
+            injectedLayer = {
+              ...commonProps,
+              ignoreGlobalFilters: persistedLayer.ignoreGlobalFilters,
+              indexPatternId: indexPatternIdFromReferences,
+              annotations: cloneDeep(persistedLayer.annotations),
+            };
+          }
+        }
+
+        return injectedLayer;
+      })
+      .filter(nonNullable),
   };
 }
 
-function getIndexPatternIdFromInitialContext(
-  layer: XYAnnotationLayerConfig,
-  initialContext?: VisualizeFieldContext | VisualizeEditorContext
-) {
-  if (initialContext && 'isVisualizeAction' in initialContext) {
-    return layer && 'indexPatternId' in layer ? layer.indexPatternId : undefined;
-  }
-}
+// TODO - was this doing anything?
+// function getIndexPatternIdFromInitialContext(
```

Comment:
:D

Reply:
Lol, forgot about this... it looks like it was added as part of https://github.com/elastic/kibana/pull/140718, but I don't see any regression in the TSVB to Lens. Would be nice if someone could double-check me on this...

Reply:
I can't find none either, let's clean it up!